src/sentry/sentry_metrics/indexer/cloudspanner/cloudspanner.py

```diff
@@ -60,28 +118,315 @@ def validate(self) -> None:
+        key_results: KeyResults,
+    ) -> None:
+        """
+        Insert the records in a transaction while handling collisions that
+        might happen during an INSERT. Collisions can happen  when multiple
+        consumers try to write records with same organization_id and string.
+
+        The basic logic is to retry the below in a loop until the transaction
+        succeeds.
+        The logic within a transaction is as follows:
+        1. Get records from the database for the organization_id and string
+        which are present in the rows.
+        2. Get records from the database for the id which are present in the rows.
+        3. Create a new list of records to be inserted which excludes the
+        records which are present in the above two lists.
+        4. Insert the new list of records in a transaction.
+
+        Once the transaction succeeds, we update the key_results by
+        performing a read of the organization_id and string which were
+        supposed to be originally inserted.
+        """
+
+        def insert_rw_transaction_uow(
+            transaction: Any, rows: Sequence[SpannerIndexerModel]
+        ) -> None:
+            existing_org_string_results = transaction.read(
+                table=self._get_table_name(use_case_id),
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=[[row.organization_id, row.string] for row in rows]),
+                index=self._get_unique_org_string_index_name(use_case_id),
+            )
+
+            existing_id_results = transaction.read(
+                table=self._get_table_name(use_case_id),
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=[[row.id] for row in rows]),
+            )
+
+            existing_records = list(existing_org_string_results) + list(existing_id_results)
+            exisiting_org_string_set = {(row[0], row[1]) for row in existing_records}
+            existing_id_string_set = {row[2] for row in existing_records}
+
+            missing_records = []
+            for row in rows:
+                if (
+                    row.organization_id,
+                    row.string,
+                ) not in exisiting_org_string_set and row.id not in existing_id_string_set:
+                    missing_records.append(row)
+
+            if missing_records:
+                transaction.insert(
+                    self._get_table_name(use_case_id),
+                    columns=get_column_names(),
+                    values=missing_records,
+                )
+
+        metrics.incr(
+            _INDEXER_DB_INSERT_METRIC,
+            tags={"batch": "false"},
+        )
+        transaction_succeeded = False
+        while not transaction_succeeded:
+            try:
+                self.database.run_in_transaction(insert_rw_transaction_uow, rows_to_insert)
+                transaction_succeeded = True
```

Comment:
Isn't spanner supposed to be retrying the transaction on its own ?

Reply:
The client library only retries for Aborted errors. But raises exceptions for others https://googleapis.dev/python/spanner/latest/transaction-usage.html#using-run-in-transaction